File: masters-thesis-radar/research-factory-v4-2026-07-27/companions/thesis-v4-mutation-tests/evaluate_mutation_corpus.mjs
#!/usr/bin/env node
/*
 * Independent JavaScript implementation of LABELLING-SPEC.md v1.0.
 *
 * Inputs are limited to the labeller-facing specification packet. This file
 * imports no research-program evaluator code and contains no reference labels.
 */

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));
const packetPath = path.join(here, "mutation-packet.json");
const outputPath = path.join(here, "js-results.json");
const layerPath = path.join(here, "js-layer-results.json");

function canonical(value, profileFloatConvention = false, parentKey = "") {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error("non-finite JSON number");
    if (
      profileFloatConvention &&
      Number.isInteger(value) &&
      (parentKey === "severity_grid" || parentKey === "zero_anchor")
    ) {
      return `${value}.0`;
    }
    return JSON.stringify(value);
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return `[${value
      .map((item) => canonical(item, profileFloatConvention, parentKey))
      .join(",")}]`;
  }
  const keys = Object.keys(value).sort();
  return `{${keys
    .map(
      (key) =>
        `${JSON.stringify(key)}:${canonical(
          value[key],
          profileFloatConvention,
          key,
        )}`,
    )
    .join(",")}}`;
}

function unsigned(object) {
  const copy = structuredClone(object);
  delete copy.signature;
  return copy;
}

function digestObject(object, profileFloatConvention = false) {
  return `sha256:${crypto
    .createHash("sha256")
    .update(canonical(unsigned(object), profileFloatConvention), "utf8")
    .digest("hex")}`;
}

function publicKeyFromRaw(base64) {
  const raw = Buffer.from(base64, "base64");
  if (raw.length !== 32) throw new Error(`Ed25519 key length ${raw.length}`);
  const spkiPrefix = Buffer.from("302a300506032b6570032100", "hex");
  return crypto.createPublicKey({
    key: Buffer.concat([spkiPrefix, raw]),
    format: "der",
    type: "spki",
  });
}

function validSignature(object, publicKey) {
  if (
    object === null ||
    typeof object !== "object" ||
    typeof object.signature !== "string"
  ) {
    return false;
  }
  let signature;
  try {
    signature = Buffer.from(object.signature, "base64");
  } catch {
    return false;
  }
  if (signature.length !== 64) return false;
  try {
    return crypto.verify(
      null,
      Buffer.from(canonical(unsigned(object)), "utf8"),
      publicKey,
      signature,
    );
  } catch {
    return false;
  }
}

function inInclusiveWindow(value, lower, upper) {
  return (
    typeof value === "string" &&
    typeof lower === "string" &&
    typeof upper === "string" &&
    lower <= value &&
    value <= upper
  );
}

function policyLayer(policy, shared, publicKey) {
  const anchor = shared.policy_anchor;
  const decisionTime = shared.decision_time;
  if (policy === null) return ["MISSING", "policy.missing", "P1"];
  if (!validSignature(policy, publicKey)) {
    return ["INVALID_SIGNATURE", "policy.signature", "P2"];
  }
  if (policy.policy_id !== anchor.required_policy_id) {
    return ["SUBSTITUTED", "policy.substituted", "P3"];
  }
  if (
    policy.version !== anchor.required_policy_version ||
    !inInclusiveWindow(decisionTime, policy.valid_from, policy.valid_until)
  ) {
    return ["STALE", "policy.stale", "P4"];
  }
  if (digestObject(policy) !== anchor.required_policy_digest) {
    return ["SUBSTITUTED", "policy.substituted", "P5"];
  }
  return ["PASS", null, "P1-P5"];
}

function evidenceLayer(evidence, shared, publicKey) {
  if (!validSignature(evidence, publicKey)) {
    return ["TAMPERED", "evidence.integrity", "E1"];
  }
  if (shared.preseen_nonces.includes(evidence.nonce)) {
    return ["REPLAY", "evidence.replay", "E2"];
  }
  return ["PASS", null, "E1-E2"];
}

function stateLayer(state, decisionTime) {
  if (state?.status !== "affirming") {
    return ["CONTRAINDICATED", "state.contraindicated", "S1"];
  }
  if (
    !inInclusiveWindow(decisionTime, state.issued_at, state.expires_at)
  ) {
    return ["STALE", "state.stale", "S2"];
  }
  if (
    typeof state.reference_digest !== "string" ||
    state.reference_digest.length === 0 ||
    state.observed_digest !== state.reference_digest
  ) {
    return ["REFERENCE_MISMATCH", "state.reference", "S3"];
  }
  return ["PASS", null, "S1-S3"];
}

function isSubset(child, parent) {
  if (!Array.isArray(child) || !Array.isArray(parent)) return false;
  const parentSet = new Set(parent.map((item) => canonical(item)));
  return child.every((item) => parentSet.has(canonical(item)));
}

function authorityLayer(authority, shared) {
  if (authority.protocol_valid !== true) {
    return ["DENY", "protocol.valid", "A0"];
  }
  let principal = shared.root_principal;
  let role = shared.root_role;
  let scope = structuredClone(shared.root_scope);
  for (let index = 0; index < authority.edges.length; index += 1) {
    const edge = authority.edges[index];
    if (edge.delegator !== principal || edge.from_role !== role) {
      return ["DENY", "edge.lineage", `A1:${index}`];
    }
    if (edge.issuer_signature_valid !== true) {
      return ["DENY", "edge.signature", `A2:${index}`];
    }
    if (index > 0 && scope.delegable !== true) {
      return ["DENY", "edge.parent_delegable", `A3:${index}`];
    }
    const transitions = shared.allowed_role_transitions[role] ?? [];
    if (!transitions.includes(edge.to_role)) {
      return ["DENY", "edge.role_transition", `A4:${index}`];
    }
    if (edge.status !== "active") {
      return ["DENY", "edge.status", `A5:${index}`];
    }
    if (
      !inInclusiveWindow(
        shared.effect_time,
        edge.not_before,
        edge.expires_at,
      )
    ) {
      return ["DENY", "edge.freshness", `A6:${index}`];
    }
    for (const field of [
      "operations",
      "resources",
      "tools",
      "audiences",
      "purposes",
    ]) {
      if (!isSubset(edge.grant[field], scope[field])) {
        return ["DENY", `edge.scope.${field}`, `A7:${index}`];
      }
    }
    if (edge.grant.currency !== scope.currency) {
      return ["DENY", "edge.scope.currency", `A8:${index}`];
    }
    if (
      typeof edge.grant.amount_max !== "number" ||
      typeof scope.amount_max !== "number" ||
      edge.grant.amount_max > scope.amount_max
    ) {
      return ["DENY", "edge.scope.amount_max", `A9:${index}`];
    }
    principal = edge.delegate;
    role = edge.to_role;
    scope = structuredClone(edge.grant);
  }

  const action = authority.action;
  if (action.subject !== principal || action.role !== role) {
    return ["DENY", "action.subject_role", "A10"];
  }
  for (const [field, gate] of [
    ["operation", "action.operation"],
    ["resource", "action.resource"],
    ["tool", "action.tool"],
    ["audience", "action.audience"],
    ["purpose", "action.purpose"],
  ]) {
    const plural = `${field}s`;
    if (!Array.isArray(scope[plural]) || !scope[plural].includes(action[field])) {
      return ["DENY", gate, `A${11 + ["operation", "resource", "tool", "audience", "purpose"].indexOf(field)}`];
    }
  }
  if (
    action.currency !== scope.currency ||
    typeof action.amount !== "number" ||
    action.amount > scope.amount_max
  ) {
    return ["DENY", "action.amount_currency", "A16"];
  }
  const receipt = authority.receipt;
  if (receipt.native_evidence_valid !== true) {
    return ["DENY", "effect.native_validity", "A17"];
  }
  if (receipt.effect_time !== shared.effect_time) {
    return ["DENY", "effect.time_binding", "A18"];
  }
  if (canonical(receipt.action) !== canonical(action)) {
    return ["DENY", "effect.action_binding", "A19"];
  }
  return ["ALLOW", null, "A0-A19"];
}

function flattenMean(rows) {
  let total = 0;
  let count = 0;
  for (const row of rows) {
    if (Array.isArray(row)) {
      for (const value of row) {
        total += value;
        count += 1;
      }
    } else {
      total += row;
      count += 1;
    }
  }
  return total / count;
}

function xorshift32(seed) {
  let state = seed >>> 0;
  return () => {
    state ^= state << 13;
    state ^= state >>> 17;
    state ^= state << 5;
    return (state >>> 0) / 4294967296;
  };
}

function quantileLinear(sorted, alpha) {
  const position = (sorted.length - 1) * alpha;
  const lower = Math.floor(position);
  const upper = Math.ceil(position);
  if (lower === upper) return sorted[lower];
  const fraction = position - lower;
  return sorted[lower] * (1 - fraction) + sorted[upper] * fraction;
}

function bootstrapLcb(rows, replicates, seed, alpha) {
  const rng = xorshift32(seed);
  const n = rows.length;
  const means = [];
  for (let replicate = 0; replicate < replicates; replicate += 1) {
    const sampled = [];
    for (let index = 0; index < n; index += 1) {
      sampled.push(rows[Math.floor(rng() * n)]);
    }
    means.push(flattenMean(sampled));
  }
  means.sort((left, right) => left - right);
  return quantileLinear(means, alpha);
}

function measurementLayer(measurement, shared, cache) {
  const profileDigest = digestObject(measurement.profile, true);
  if (profileDigest !== measurement.required_profile_digest) {
    return ["PROFILE_MISMATCH", "measurement.profile", "M1", {
      profile_digest: profileDigest,
    }];
  }
  const rows = shared.datasets[measurement.dataset];
  if (!Array.isArray(rows)) {
    throw new Error(`missing dataset ${measurement.dataset}`);
  }
  if (!cache.has(measurement.dataset)) {
    const point = flattenMean(rows);
    const lcb = bootstrapLcb(
      rows,
      shared.bootstrap.replicates,
      shared.bootstrap.seed_by_dataset[measurement.dataset],
      measurement.alpha,
    );
    cache.set(measurement.dataset, { point, lcb });
  }
  const values = cache.get(measurement.dataset);
  if (values.point < measurement.threshold) {
    return ["FAIL_POINT", "measurement.point", "M2", values];
  }
  if (values.lcb < measurement.threshold) {
    return ["FAIL_LCB", "measurement.confidence", "M3", values];
  }
  return ["PASS", null, "M1-M3", values];
}

function evaluate(transaction, shared, publicKey, measurementCache) {
  const layers = {
    policy: policyLayer(
      transaction.policy,
      shared.policy_evidence,
      publicKey,
    ),
    evidence: evidenceLayer(
      transaction.evidence,
      shared.policy_evidence,
      publicKey,
    ),
    state: stateLayer(
      transaction.attestation_result,
      shared.state.decision_time,
    ),
    authority: authorityLayer(transaction.authority, shared.authority),
    measurement: measurementLayer(
      transaction.measurement,
      shared.measurement,
      measurementCache,
    ),
  };
  const order = ["policy", "evidence", "state", "authority", "measurement"];
  const passing = {
    policy: "PASS",
    evidence: "PASS",
    state: "PASS",
    authority: "ALLOW",
    measurement: "PASS",
  };
  const first = order.find((name) => layers[name][0] !== passing[name]);
  return {
    transaction_id: transaction.transaction_id,
    verdict: first === undefined ? "ALLOW" : "DENY",
    first_rejecting_gate:
      first === undefined ? "verified" : layers[first][1],
    layers: Object.fromEntries(
      order.map((name) => [
        name,
        {
          result: layers[name][0],
          gate: layers[name][1],
          rule: layers[name][2],
          diagnostics: layers[name][3] ?? null,
        },
      ]),
    ),
  };
}

const packet = JSON.parse(fs.readFileSync(packetPath, "utf8"));
const publicKey = publicKeyFromRaw(
  packet.shared_inputs.policy_evidence.public_key_raw_b64,
);
const measurementCache = new Map();
const evaluations = packet.transactions.map((transaction) =>
  evaluate(transaction, packet.shared_inputs, publicKey, measurementCache),
);
if (evaluations.length !== packet.transactions.length) {
  throw new Error(
    `expected ${packet.transactions.length} transactions, got ${evaluations.length}`,
  );
}
if (
  new Set(evaluations.map((item) => item.transaction_id)).size !==
  evaluations.length
) {
  throw new Error("transaction IDs are not unique");
}

const answers = evaluations.map((item) => ({
  transaction_id: item.transaction_id,
  labeller_id: "mutation-js-v1",
  verdict: item.verdict,
  first_rejecting_gate: item.first_rejecting_gate,
  confidence: 5,
  rationale:
    item.verdict === "ALLOW"
      ? "All five typed layers pass under LABELLING-SPEC.md v1.0."
      : `The first failing layer in fixed composition order rejects at ${item.first_rejecting_gate}.`,
}));

fs.writeFileSync(outputPath, `${JSON.stringify(answers, null, 2)}\n`);
fs.writeFileSync(layerPath, `${JSON.stringify(evaluations, null, 2)}\n`);

const summary = {
  implementation: "mutation-js-v1",
  transactions: answers.length,
  allows: answers.filter((item) => item.verdict === "ALLOW").length,
  denies: answers.filter((item) => item.verdict === "DENY").length,
  gate_counts: Object.fromEntries(
    [...new Set(answers.map((item) => item.first_rejecting_gate))]
      .sort()
      .map((gate) => [
        gate,
        answers.filter((item) => item.first_rejecting_gate === gate).length,
      ]),
  ),
  measurement_diagnostics: Object.fromEntries(measurementCache),
};
process.stdout.write(`${JSON.stringify(summary, null, 2)}\n`);

File: masters-thesis-radar/research-factory-v4-2026-07-27/companions/mutation-tests/evaluate_mutation_corpus.py
#!/usr/bin/env python3
"""Independent Python implementation of LABELLING-SPEC.md for mutations.

This evaluator reads only mutation-packet.json. It does not read the mutation
oracle, sealed labels, or research-program evaluator source.
"""

from __future__ import annotations

import base64
import copy
import hashlib
import json
from pathlib import Path
from typing import Any

import numpy as np
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


HERE = Path(__file__).resolve().parent


def canonical(value: Any) -> str:
    return json.dumps(
        value, ensure_ascii=False, sort_keys=True, separators=(",", ":")
    )


def unsigned(value: dict[str, Any]) -> dict[str, Any]:
    result = copy.deepcopy(value)
    result.pop("signature", None)
    return result


def digest_object(value: dict[str, Any]) -> str:
    return "sha256:" + hashlib.sha256(
        canonical(unsigned(value)).encode("utf-8")
    ).hexdigest()


def valid_signature(
    value: Any, public_key: Ed25519PublicKey
) -> bool:
    if not isinstance(value, dict) or not isinstance(value.get("signature"), str):
        return False
    try:
        signature = base64.b64decode(value["signature"], validate=True)
        if len(signature) != 64:
            return False
        public_key.verify(signature, canonical(unsigned(value)).encode("utf-8"))
        return True
    except Exception:
        return False


def in_window(value: Any, lower: Any, upper: Any) -> bool:
    return (
        isinstance(value, str)
        and isinstance(lower, str)
        and isinstance(upper, str)
        and lower <= value <= upper
    )


def policy_layer(
    policy: Any, shared: dict[str, Any], public_key: Ed25519PublicKey
) -> tuple[str, str | None, str, dict[str, Any] | None]:
    anchor = shared["policy_anchor"]
    decision_time = shared["decision_time"]
    if policy is None:
        return ("MISSING", "policy.missing", "P1", None)
    if not valid_signature(policy, public_key):
        return ("INVALID_SIGNATURE", "policy.signature", "P2", None)
    if policy.get("policy_id") != anchor["required_policy_id"]:
        return ("SUBSTITUTED", "policy.substituted", "P3", None)
    if policy.get("version") != anchor["required_policy_version"] or not in_window(
        decision_time, policy.get("valid_from"), policy.get("valid_until")
    ):
        return ("STALE", "policy.stale", "P4", None)
    if digest_object(policy) != anchor["required_policy_digest"]:
        return ("SUBSTITUTED", "policy.substituted", "P5", None)
    return ("PASS", None, "P1-P5", None)


def evidence_layer(
    evidence: Any, shared: dict[str, Any], public_key: Ed25519PublicKey
) -> tuple[str, str | None, str, dict[str, Any] | None]:
    if not valid_signature(evidence, public_key):
        return ("TAMPERED", "evidence.integrity", "E1", None)
    if evidence.get("nonce") in shared["preseen_nonces"]:
        return ("REPLAY", "evidence.replay", "E2", None)
    return ("PASS", None, "E1-E2", None)


def state_layer(
    state: Any, decision_time: str
) -> tuple[str, str | None, str, dict[str, Any] | None]:
    if not isinstance(state, dict) or state.get("status") != "affirming":
        return ("CONTRAINDICATED", "state.contraindicated", "S1", None)
    if not in_window(decision_time, state.get("issued_at"), state.get("expires_at")):
        return ("STALE", "state.stale", "S2", None)
    if (
        not isinstance(state.get("reference_digest"), str)
        or not state["reference_digest"]
        or state.get("observed_digest") != state["reference_digest"]
    ):
        return ("REFERENCE_MISMATCH", "state.reference", "S3", None)
    return ("PASS", None, "S1-S3", None)


def is_subset(child: Any, parent: Any) -> bool:
    if not isinstance(child, list) or not isinstance(parent, list):
        return False
    parent_values = {canonical(item) for item in parent}
    return all(canonical(item) in parent_values for item in child)


def authority_layer(
    authority: dict[str, Any], shared: dict[str, Any]
) -> tuple[str, str | None, str, dict[str, Any] | None]:
    if authority.get("protocol_valid") is not True:
        return ("DENY", "protocol.valid", "A0", None)

    principal = shared["root_principal"]
    role = shared["root_role"]
    scope = copy.deepcopy(shared["root_scope"])
    for index, edge in enumerate(authority["edges"]):
        if edge.get("delegator") != principal or edge.get("from_role") != role:
            return ("DENY", "edge.lineage", f"A1:{index}", None)
        if edge.get("issuer_signature_valid") is not True:
            return ("DENY", "edge.signature", f"A2:{index}", None)
        if index > 0 and scope.get("delegable") is not True:
            return ("DENY", "edge.parent_delegable", f"A3:{index}", None)
        if edge.get("to_role") not in shared["allowed_role_transitions"].get(
            role, []
        ):
            return ("DENY", "edge.role_transition", f"A4:{index}", None)
        if edge.get("status") != "active":
            return ("DENY", "edge.status", f"A5:{index}", None)
        if not in_window(
            shared["effect_time"], edge.get("not_before"), edge.get("expires_at")
        ):
            return ("DENY", "edge.freshness", f"A6:{index}", None)
        for field in (
            "operations",
            "resources",
            "tools",
            "audiences",
            "purposes",
        ):
            if not is_subset(edge["grant"].get(field), scope.get(field)):
                return (
                    "DENY",
                    f"edge.scope.{field}",
                    f"A7:{index}",
                    None,
                )
        if edge["grant"].get("currency") != scope.get("currency"):
            return ("DENY", "edge.scope.currency", f"A8:{index}", None)
        amount = edge["grant"].get("amount_max")
        if (
            not isinstance(amount, (int, float))
            or isinstance(amount, bool)
            or not isinstance(scope.get("amount_max"), (int, float))
            or amount > scope["amount_max"]
        ):
            return ("DENY", "edge.scope.amount_max", f"A9:{index}", None)
        principal = edge["delegate"]
        role = edge["to_role"]
        scope = copy.deepcopy(edge["grant"])

    action = authority["action"]
    if action.get("subject") != principal or action.get("role") != role:
        return ("DENY", "action.subject_role", "A10", None)
    action_fields = (
        ("operation", "operations", "A11"),
        ("resource", "resources", "A12"),
        ("tool", "tools", "A13"),
        ("audience", "audiences", "A14"),
        ("purpose", "purposes", "A15"),
    )
    for singular, plural, rule in action_fields:
        if action.get(singular) not in scope.get(plural, []):
            return ("DENY", f"action.{singular}", rule, None)
    if (
        action.get("currency") != scope.get("currency")
        or not isinstance(action.get("amount"), (int, float))
        or isinstance(action.get("amount"), bool)
        or action["amount"] > scope["amount_max"]
    ):
        return ("DENY", "action.amount_currency", "A16", None)

    receipt = authority["receipt"]
    if receipt.get("native_evidence_valid") is not True:
        return ("DENY", "effect.native_validity", "A17", None)
    if receipt.get("effect_time") != shared["effect_time"]:
        return ("DENY", "effect.time_binding", "A18", None)
    if canonical(receipt.get("action")) != canonical(action):
        return ("DENY", "effect.action_binding", "A19", None)
    return ("ALLOW", None, "A0-A19", None)


def flatten_mean(rows: list[Any]) -> float:
    return float(np.asarray(rows, dtype=float).mean())


def bootstrap_lcb(
    rows: list[Any], replicates: int, seed: int, alpha: float
) -> float:
    array = np.asarray(rows, dtype=float)
    rng = np.random.default_rng(seed)
    indices = rng.integers(0, len(array), size=(replicates, len(array)))
    replicate_means = array[indices].mean(axis=tuple(range(1, array.ndim + 1)))
    return float(np.quantile(replicate_means, alpha, method="linear"))


def measurement_layer(
    measurement: dict[str, Any],
    shared: dict[str, Any],
    cache: dict[tuple[str, float], dict[str, float]],
) -> tuple[str, str | None, str, dict[str, Any] | None]:
    profile_digest = digest_object(measurement["profile"])
    if profile_digest != measurement["required_profile_digest"]:
        return (
            "PROFILE_MISMATCH",
            "measurement.profile",
            "M1",
            {"profile_digest": profile_digest},
        )
    rows = shared["datasets"].get(measurement["dataset"])
    if not isinstance(rows, list):
        raise RuntimeError(f"missing dataset {measurement['dataset']}")
    key = (measurement["dataset"], measurement["alpha"])
    if key not in cache:
        cache[key] = {
            "point": flatten_mean(rows),
            "lcb": bootstrap_lcb(
                rows,
                shared["bootstrap"]["replicates"],
                shared["bootstrap"]["seed_by_dataset"][measurement["dataset"]],
                measurement["alpha"],
            ),
        }
    values = cache[key]
    if values["point"] < measurement["threshold"]:
        return ("FAIL_POINT", "measurement.point", "M2", values)
    if values["lcb"] < measurement["threshold"]:
        return ("FAIL_LCB", "measurement.confidence", "M3", values)
    return ("PASS", None, "M1-M3", values)


def evaluate(
    transaction: dict[str, Any],
    shared: dict[str, Any],
    public_key: Ed25519PublicKey,
    measurement_cache: dict[tuple[str, float], dict[str, float]],
) -> dict[str, Any]:
    layer_values = {
        "policy": policy_layer(
            transaction["policy"], shared["policy_evidence"], public_key
        ),
        "evidence": evidence_layer(
            transaction["evidence"], shared["policy_evidence"], public_key
        ),
        "state": state_layer(
            transaction["attestation_result"], shared["state"]["decision_time"]
        ),
        "authority": authority_layer(transaction["authority"], shared["authority"]),
        "measurement": measurement_layer(
            transaction["measurement"], shared["measurement"], measurement_cache
        ),
    }
    order = ("policy", "evidence", "state", "authority", "measurement")
    passing = {
        "policy": "PASS",
        "evidence": "PASS",
        "state": "PASS",
        "authority": "ALLOW",
        "measurement": "PASS",
    }
    first = next(
        (name for name in order if layer_values[name][0] != passing[name]), None
    )
    layers = {
        name: {
            "result": layer_values[name][0],
            "gate": layer_values[name][1],
            "rule": layer_values[name][2],
            "diagnostics": layer_values[name][3],
        }
        for name in order
    }
    return {
        "transaction_id": transaction["transaction_id"],
        "verdict": "ALLOW" if first is None else "DENY",
        "first_rejecting_gate": "verified" if first is None else layers[first]["gate"],
        "layers": layers,
    }


packet = json.loads((HERE / "mutation-packet.json").read_text(encoding="utf-8"))
raw_key = base64.b64decode(
    packet["shared_inputs"]["policy_evidence"]["public_key_raw_b64"],
    validate=True,
)
public_key = Ed25519PublicKey.from_public_bytes(raw_key)
measurement_cache: dict[tuple[str, float], dict[str, float]] = {}
evaluations = [
    evaluate(item, packet["shared_inputs"], public_key, measurement_cache)
    for item in packet["transactions"]
]

if len(evaluations) != len(packet["transactions"]):
    raise RuntimeError("evaluation count mismatch")
if len({row["transaction_id"] for row in evaluations}) != len(evaluations):
    raise RuntimeError("transaction IDs are not unique")

answers = [
    {
        "transaction_id": row["transaction_id"],
        "labeller_id": "mutation-python-v1",
        "verdict": row["verdict"],
        "first_rejecting_gate": row["first_rejecting_gate"],
        "confidence": 5,
        "rationale": (
            "All five typed layers pass under LABELLING-SPEC.md v1.0."
            if row["verdict"] == "ALLOW"
            else "The first failing layer in fixed composition order rejects at "
            + row["first_rejecting_gate"]
            + "."
        ),
    }
    for row in evaluations
]

(HERE / "python-results.json").write_text(
    json.dumps(answers, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
)
(HERE / "python-layer-results.json").write_text(
    json.dumps(evaluations, ensure_ascii=False, indent=2) + "\n",
    encoding="utf-8",
)

print(
    json.dumps(
        {
            "implementation": "mutation-python-v1",
            "transactions": len(answers),
            "allows": sum(row["verdict"] == "ALLOW" for row in answers),
            "denies": sum(row["verdict"] == "DENY" for row in answers),
            "measurement_diagnostics": {
                f"{dataset}@{alpha}": values
                for (dataset, alpha), values in measurement_cache.items()
            },
        },
        indent=2,
        sort_keys=True,
    )
)
